hail/src/main/scala/is/hail/expr/ir/IR.scala

```diff
@@ -117,10 +117,22 @@ object If {
     if (cnsq.typ == altr.typ)
       If(cond, cnsq, altr)
     else {
-      val t = unifyType.getOrElse(cnsq.typ.deepOptional())
-      If(cond,
-        PruneDeadFields.upcast(cnsq, t),
-        PruneDeadFields.upcast(altr, t))
+      cnsq match {
+        case NA(_) => If(cond, NA(altr.typ), altr)
+        case Die(msg, _) => If(cond, Die(msg, altr.typ), altr)
+        case Literal(_, value) => If(cond, Literal(altr.typ, value), altr)
```

Comment:
Should this check that `Literal(altr.typ, value)` is well-typed?

Reply:
hmm. Yeah, probably.